docs/mypy_plugin.md

```diff
@@ -0,0 +1,137 @@
+Pydantic works well with [mypy](http://mypy-lang.org/) right [out of the box](usage/mypy.md).
+
+However, Pydantic also ships with a mypy plugin that adds a number of important pydantic-specific
+features to mypy that improve its ability to type-check your code.
+
+For example, consider the following script:
+```py
+{!.tmp_examples/mypy.py!}
+```
+
+Without any set up, mypy catches one of the errors:
+```
+13: error: "Model" has no attribute "middle_name"
+```
+
+But with the plugin enabled, it catches both:
+```
+13: error: "Model" has no attribute "middle_name"
+16: error: Missing named argument "age" for "Model"
+16: error: Missing named argument "list_of_ints" for "Model"
+```
+
+With the pydantic mypy plugin, you can fearlessly refactor your models knowing mypy will catch any mistakes
+if your field names or types change.
+
+There are other benefits too! See below for more details.
+
+### Plugin Capabilities
+
+#### Generate a signature for `Model.__init__`
+* Any required fields that don't have dynamically-determined aliases will be included as required
+  keyword arguments.
+* If `Config.allow_population_by_field_name=True`, the generated signature will use the field names,
```

Comment:
The way mypy works in this context is that you can basically programatically generate a method definition, and then it is type-checked as though it had been written in code. So, you can't really encode the alias-handling logic in a way mypy supports -- it has to be checked like a normal python method.

I can't imagine a circumstance under which you would have `allow_population_by_field_name = True`, and also prefer to use the *alias* in your code. It seems to me the vast majority of usage of this feature is to have a pythonic name for use in code, and a non-pythonic (e.g., camel case) alias.

I'm not saying it's not possible that someone would prefer to use the alias in their code, but then I just don't understand why they wouldn't have named the variable that.

Reply:
We could set it up so that whenever mypy can determine the alias, that is what is used, but I think that is also problematic: for example, how should we handle cases where the alias contains a hyphen? In the end, it seems to me that the 1) highest-consistency, 2) highest user-convenience, and 3) easiest to implement solution is to just use the non-alias names whenever that is an option.

Reply:
Theoretically, I guess we could create overloaded signatures with both versions of the argument where possible, but that would lead to 2^n overloads where n is the number of aliases, so the error messages would get completely out of control.

Reply:
I was just thinking of creating a signature something like:

```py
def __init__(self, foo_bar: str = None, fooBar: str = None, user_id: int = None, userId: int = None):
    ...
```
But as i wrote that I realised it wouldn't work for required fields.

Let's stick to the current behaviour.